File: gee/checa_area.py
"""Diagnostico de uma area candidata antes de fixar a area de estudo.

Reporta quantas quadriculas MGRS cobrem a area (mais de uma = faixa de
sobreposicao, gera observacoes redundantes por passagem), quantas orbitas, a
densidade de imagens por ano e quantas observacoes sobrevivem a mascara de nuvem.

    python gee/checa_area.py
    python gee/checa_area.py --area area/candidata_02.geojson
"""

from __future__ import annotations

import argparse
import collections
import datetime as dt
import pathlib
import sys

from comum import autentica, carrega_area, cfg_data, colecao_ndvi, le_config


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--area", help="GeoJSON alternativo, para comparar candidatas")
    args = ap.parse_args()

    import ee

    cfg = le_config()
    if args.area:
        cfg["area_geojson"] = args.area
    autentica()

    area = carrega_area(cfg)
    inicio, fim = cfg_data(cfg, "data_inicio"), cfg_data(cfg, "data_fim")

    print(f"\narea    : {cfg['area_geojson']}")
    ha = area.area(maxError=1).getInfo() / 10_000.0
    print(f"tamanho : {ha:.1f} ha  (~{int(ha * 100)} pixels de 10 m)")

    base = (ee.ImageCollection(cfg["colecao"])
            .filterBounds(area)
            .filterDate(str(inicio), str(fim)))

    n = base.size().getInfo()
    print(f"periodo : {inicio} a {fim}  ->  {n} imagens brutas")
    if n == 0:
        print("\nNenhuma imagem. Verifique o poligono e as datas.")
        return 1

    tiles = sorted(set(base.aggregate_array("MGRS_TILE").getInfo()))
    orbitas = sorted(set(base.aggregate_array("SENSING_ORBIT_NUMBER").getInfo()))

    print("\n--- cobertura ---")
    print(f"quadriculas MGRS : {', '.join(tiles)}")
    print(f"orbitas          : {', '.join(str(o) for o in orbitas)}")

    if len(tiles) > 1:
        print(f"\n  {len(tiles)} quadriculas: a area esta na faixa de sobreposicao")
        print("  e cada passagem entrega o mesmo pixel em dois produtos.")

    if len(orbitas) > 1:
        print(f"\n  {len(orbitas)} orbitas: mais revisita, geometria de visada variavel.")

    datas = base.aggregate_array("system:time_start").getInfo()
    por_ano = collections.Counter(
        dt.datetime.fromtimestamp(t / 1000, dt.timezone.utc).year for t in datas)

    col = colecao_ndvi(cfg, area, inicio, fim)
    stats = col.select("NDVI").reduce(ee.Reducer.count()).reduceRegion(
        reducer=ee.Reducer.mean().combine(ee.Reducer.minMax(), sharedInputs=True),
        geometry=area, scale=int(cfg["escala"]), maxPixels=1e9).getInfo()

    print("\n--- densidade temporal ---")
    print("imagens brutas por ano:")
    for ano in sorted(por_ano):
        print(f"  {ano}: {por_ano[ano]:>4}")

    media = stats.get("NDVI_count_mean")
    if media is not None:
        anos = (fim - inicio).days / 365.25
        print(f"\nobservacoes validas por pixel (limiar {cfg['limiar_nuvem']}):")
        print(f"  media : {media:.0f}   ({media / anos:.0f} por ano)")
        print(f"  min   : {stats.get('NDVI_count_min'):.0f}")
        print(f"  max   : {stats.get('NDVI_count_max'):.0f}")
        print(f"  aproveitamento: {100 * media / n:.0f}% das passagens")
        if media < 60:
            print("\n  poucas observacoes por pixel: historico curto compromete o")
            print("  ajuste sazonal.")

    print("\n--- criterios ---")
    print(f"  quadricula unica            : {'sim' if len(tiles) == 1 else 'nao'}")
    print(f"  >= 60 observacoes por pixel : {'sim' if (media or 0) >= 60 else 'nao'}")
    print("  confirmar visualmente com gee/inspeciona_area.js\n")
    return 0


if __name__ == "__main__":
    sys.path.insert(0, str(pathlib.Path(__file__).parent))
    raise SystemExit(main())

File: inspecao/gee/checa_area.js
// Diagnostico de uma area candidata: quadriculas, orbitas e observacoes validas.
//
//   node inspecao/gee/checa_area.js
//   node inspecao/gee/checa_area.js --area area/candidata_02.geojson

import { parseArgs } from 'node:util';
import { autentica, avalia, carregaArea, cfgData, colecaoNdvi, ee, leConfig } from '../../gee/comum.js';

const { values: args } = parseArgs({ options: { area: { type: 'string' } } });

const cfg = leConfig();
if (args.area) cfg.area_geojson = args.area;
await autentica();

const area = carregaArea(cfg);
const inicio = cfgData(cfg, 'data_inicio');
const fim = cfgData(cfg, 'data_fim');

console.log(`\narea    : ${cfg.area_geojson}`);
const ha = (await avalia(area.area({ maxError: 1 }))) / 10000;
console.log(`tamanho : ${ha.toFixed(1)} ha  (~${Math.round(ha * 100)} pixels de 10 m)`);

const base = ee.ImageCollection(cfg.colecao).filterBounds(area).filterDate(inicio, fim);
const n = await avalia(base.size());
console.log(`periodo : ${inicio} a ${fim}  ->  ${n} imagens brutas`);
if (n === 0) {
  console.log('\nNenhuma imagem. Verifique o poligono e as datas.');
  process.exit(1);
}

const tiles = [...new Set(await avalia(base.aggregate_array('MGRS_TILE')))].sort();
const orbitas = [...new Set(await avalia(base.aggregate_array('SENSING_ORBIT_NUMBER')))].sort((a, b) => a - b);

console.log('\n--- cobertura ---');
console.log(`quadriculas MGRS : ${tiles.join(', ')}`);
console.log(`orbitas          : ${orbitas.join(', ')}`);
if (tiles.length > 1) console.log(`\n  ${tiles.length} quadriculas: area na faixa de sobreposicao`);
if (orbitas.length > 1) console.log(`\n  ${orbitas.length} orbitas: geometria de visada variavel`);

const porAno = {};
for (const t of await avalia(base.aggregate_array('system:time_start'))) {
  const ano = new Date(t).getUTCFullYear();
  porAno[ano] = (porAno[ano] || 0) + 1;
}

const contagem = colecaoNdvi(cfg, area, inicio, fim).select('NDVI').reduce(ee.Reducer.count());
const stats = await avalia(contagem.reduceRegion({
  reducer: ee.Reducer.mean().combine({ reducer2: ee.Reducer.minMax(), sharedInputs: true }),
  geometry: area,
  scale: Number(cfg.escala),
  maxPixels: 1e9,
}));

console.log('\n--- densidade temporal ---');
console.log('imagens brutas por ano:');
for (const ano of Object.keys(porAno).sort()) console.log(`  ${ano}: ${String(porAno[ano]).padStart(4)}`);

const media = stats.NDVI_count_mean;
if (media != null) {
  const anos = (new Date(fim) - new Date(inicio)) / (365.25 * 86400000);
  console.log(`\nobservacoes validas por pixel (limiar ${cfg.limiar_nuvem}):`);
  console.log(`  media : ${media.toFixed(0)}   (${(media / anos).toFixed(0)} por ano)`);
  console.log(`  min   : ${stats.NDVI_count_min}`);
  console.log(`  max   : ${stats.NDVI_count_max}`);
  console.log(`  aproveitamento: ${((100 * media) / n).toFixed(0)}% das passagens`);
}

console.log('\n--- criterios ---');
console.log(`  quadricula unica            : ${tiles.length === 1 ? 'sim' : 'nao'}`);
console.log(`  >= 60 observacoes por pixel : ${(media || 0) >= 60 ? 'sim' : 'nao'}`);
console.log('  confirmar visualmente com inspecao/gee/inspeciona_area.js\n');
